consumer/src/main/java/com/flipkart/varadhi/consumer/ConcurrencyControl.java

```diff
@@ -0,0 +1,178 @@
+package com.flipkart.varadhi.consumer;
+
+import com.flipkart.varadhi.CircularQueue;
+import com.flipkart.varadhi.consumer.concurrent.Context;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+/**
+ * Not supposed to be thread safe. It is expected that tasks will be enqueued & dequeued from the same thread which
```

Comment:
In https://github.com/flipkart-incubator/varadhi/issues/113 it is mentioned as a thread safe DS. Can you clarify?

Reply:
aah, I initially thought that this DS needs to be completely thread safe. But while implementing I realised that is not the case. 
This class will see more changes in future actually, wrt `auto-recovery from failure`. So thread safety requirements may change. But now I am hoping that main methods can be simple and does not require to be thread safe. All needed sychronization is provided by some of the atomic variables.